pkg/auth/providers/oidc/oidc_provider.go

```diff
@@ -0,0 +1,418 @@
+		Provider:      Name,
+	}
+
+	principals = append(principals, p)
+	return principals, nil
+}
+
+func (o *oidcProvider) GetPrincipal(principalID string, token v3.Token) (v3.Principal, error) {
+	externalID, principalType := splitPrincipalID(principalID)
+	if externalID == "" && principalType == "" {
+		return v3.Principal{}, fmt.Errorf("[generic oidc]: invalid id %v", principalID)
+	}
+	if principalType != UserType && principalType != GroupType {
+		return v3.Principal{}, fmt.Errorf("[generic oidc]: invalid principal type")
+	}
+	p := v3.Principal{
+		ObjectMeta:  metav1.ObjectMeta{Name: principalType + "://" + externalID},
+		DisplayName: externalID,
+		LoginName:   externalID,
+		Provider:    Name,
+	}
+
+	p = o.toPrincipalFromPrincipal(principalType, p, &token)
+	return p, nil
+
+}
+
+func (o *oidcProvider) CustomizeSchema(schema *types.Schema) {
+	schema.ActionHandler = o.actionHandler
+	schema.Formatter = o.formatter
+}
+
+func (o *oidcProvider) TransformToAuthProvider(authConfig map[string]interface{}) (map[string]interface{}, error) {
+	p := common.TransformToAuthProvider(authConfig)
+	p[publicclient.OIDCProviderFieldRedirectURL] = o.getRedirectURL(authConfig)
+	return p, nil
+}
+
+func (o *oidcProvider) getRedirectURL(config map[string]interface{}) string {
+	return fmt.Sprintf(
+		"%s?client_id=%s&response_type=code&redirect_uri=%s",
+		config["authEndpoint"],
+		config["clientId"],
+		config["rancherUrl"],
+	)
+}
+
+func (o *oidcProvider) RefetchGroupPrincipals(principalID string, secret string) ([]v3.Principal, error) {
+	return nil, errors.New("[generic oidc] RefetchGroupPrincipals: Not implemented")
+}
+
+func (o *oidcProvider) CanAccessWithGroupProviders(userPrincipalID string, groupPrincipals []v3.Principal) (bool, error) {
+	config, err := o.getOIDCConfig()
+	if err != nil {
+		logrus.Errorf("[generic oidc]: error fetching OIDCConfig: %v", err)
+		return false, err
+	}
+	allowed, err := o.userMGR.CheckAccess(config.AccessMode, config.AllowedPrincipalIDs, userPrincipalID, groupPrincipals)
+	if err != nil {
+		return false, err
+	}
+	return allowed, nil
+}
+
+func (o *oidcProvider) toPrincipal(principalType string, userInfo *UserInfo, token *v3.Token) v3.Principal {
+	displayName := userInfo.Name
+	if displayName == "" {
+		displayName = userInfo.Email
+	}
+
+	p := v3.Principal{
+		ObjectMeta:    metav1.ObjectMeta{Name: Name + "_" + principalType + "://" + userInfo.Subject},
+		DisplayName:   displayName,
+		LoginName:     userInfo.Email,
+		Provider:      Name,
+		PrincipalType: principalType,
+		Me:            false,
+	}
+
+	if principalType == UserType {
+		p.PrincipalType = UserType
+		if token != nil {
+			p.Me = o.isThisUserMe(token.UserPrincipal, p)
+		}
+	} else {
+		p.PrincipalType = GroupType
+		if token != nil {
+			p.MemberOf = o.tokenMGR.IsMemberOf(*token, p)
+		}
+	}
+	return p
+}
+
+func (o *oidcProvider) toPrincipalFromPrincipal(principalType string, princ v3.Principal, token *v3.Token) v3.Principal {
```

Comment:
maybe we should name this getPrincipalFromToken?

Reply:
renamed